ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DynamicPartitionPruner.java

```diff
@@ -514,4 +569,29 @@ private boolean checkForSourceCompletion(String name) {
     }
     return false;
   }
+
+  /**
+   * Recursively replaces the ExprNodeDynamicListDesc to the list of the actual values. As a result of this call the
+   * original expression is modified so it can be used for pushing down to the TableScan for filtering the data at the
+   * source.
+   * <p>
+   * Please make sure to clone the predicate if needed since the original node will be modified.
+   * @param node The node we are traversing
+   * @param dynArgs The constant values we are substituting
+   */
+  private void replaceDynamicLists(ExprNodeDesc node, Collection<ExprNodeConstantDesc> dynArgs) {
+    List<ExprNodeDesc> children = node.getChildren();
+    if (children != null && !children.isEmpty()) {
+      ListIterator<ExprNodeDesc> iterator = node.getChildren().listIterator();
+      while (iterator.hasNext()) {
+        ExprNodeDesc child = iterator.next();
+        if (child instanceof ExprNodeDynamicListDesc) {
+          iterator.remove();
+          dynArgs.forEach(iterator::add);
```

Comment:
I strongly suspect that this method is problematic; what will happen if you have filter for 2 different columns or 2 different values?
```
a IN L1 and b IN L2
```

Reply:
This is where my knowledge might be insufficient, but we have these `SourceInfo` objects we try to evaluate.

The `SourceInfo` has the following attributes:
```
    public final ExprNodeDesc partKey;
    public final ExprNodeDesc predicate;  // <- don't mind this as this is created by me
    public final Deserializer deserializer;
    public final StructObjectInspector soi;
    public final StructField field;
    public final ObjectInspector fieldInspector;
    /* List of partitions that are required - populated from processing each event */
    public Set<Object> values = new HashSet<Object>();
    /* Whether to skipPruning - depends on the payload from an event which may signal skip - if the event payload is too large */
    public AtomicBoolean skipPruning = new AtomicBoolean();
    public final String columnName;
    public final String columnType;
    private boolean mustKeepOnePartition;
```

This suggests to me that every SourceInfo is only related to a single column. I added the `predicate` at the same place where the other fields are added, so I think the `predicate` should only contain a single column too.

Also I have run tests with `a=L1 and b=L2` conditions and found that the condition is split to multiple expressions and `SourceInfo` objects - but that was only a single test, so....

Maybe we should return false in `addDynamicSplitPruningEdge` if we have multiple columns so we can bail out early?

Reply:
Added a check for `addDynamicSplitPruningEdge` which returns `false` if there are multiple columns in the expression to filter